Doc PR: explains the numbering scheme of the PCA 10059

JPFrancoia opened the pull request:
Adding a readme for the PCA10059 board, to explain the numbering scheme of the software Pins.

This is just a documentation PR, to spare some trouble to newcomers. Feel free to close if you think it's not necessary.

esmil committed:
py, extmod: Introduce and use MP_FALLTHROUGH macro.

Newer GCC versions are able to warn about switch cases that fall
through.  This is usually a sign of a forgotten break statement, but in
the few cases where a fall through is intended we annotate it with this
macro to avoid the warning.

esmil committed:
py/vmentrytable: Ignore GCC -Woverride-init.

Like Clang, GCC warns about this file, but only with -Woverride-init
which is enabled by -Wextra. Disable the warnings for this file just
like we do for Clang to make -Wextra happy.

esmil committed:
py/objexcept: Compare mp_emergency_exception_buf_size signed.

mp_emergency_exception_buf_size is signed, so let's make sure we compare
it as such.

jimmo committed:
py/objtype: Handle __dict__ attribute when type has no locals.

robert-hh committed:
esp32/mpconfigport.h: Seed the urandom module on import.

For seeding, the RNG function of the ESP-IDF is used, which is told to be a
true RNG, at least when WiFi or Bluetooth is enabled.  Seeding on import is
as per CPython.  To obtain a reproducible sequence of pseudo-random numbers
one must explicitly seed with a known value.

robert-hh committed:
esp8266/mpconfigport.h: Seed the urandom module on import.

For seeding, the hardware RNG of the esp8266 is used.

robert-hh committed:
stm32/mpconfigport.h: Seed the urandom module on import.

For seeding the rng_get function is used, which is also the heart of
uos.urandom and pyb.rng, and is a hardware RNG where available.

stinos committed:
py/makeqstrdefs.py: Fix beaviour when scanning non-C preprocessed files.

When process_file() is passed a preprocessed C++ file for instance it won't
find any lines containing .c files and the last_fname variable remains
None, so handle that gracefully.

andrewleech committed:
unix/mpconfigport.h: Enable MICROPY_PY_DELATTR_SETATTR.

This is a generally useful feature and because it's part of the object
model it cannot be added at runtime by some loadable Python code, so enable
it on the standard unix build.

dpgeorge committed:
esp32: Use path relative to root for netutils/timeutils headers.

Signed-off-by: Damien George <damien@micropython.org>

jimmo committed:
stm32/boards/NUCLEO_WB55: Add standalone WB55 FUS/WS firmware updater.

This commit adds a script that can be run on-device to install FUS and WS
binaries from the filesystem.  Instructions for use are provided in
the rfcore_firmware.py file.

The commit also removes unneeded functionality from the existing rfcore.py
debug script (and renames it rfcore_debug.py).

jimmo committed:
stm32/rfcore: Add Python API for basic rfcore operations.

The new functions provide FUS/WS status, version and SYS HCI commands:
- stm.rfcore_status()
- stm.rfcore_fw_version(fw_id)
- stm.rfcore_sys_hci(ogf, ocf, cmd)

JPFrancoia committed:
Adding a readme for the PCA10059 board, to explain the numbering scheme
of the software Pins.

JPFrancoia committed:
Adding a readme for the PCA10059 board, to explain the numbering scheme
of the software Pins.

dpgeorge committed:
stm32/Makefile: Allow boards to extend SRC_C, SRC_O and OBJ variables.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/main: Move update_reset_mode to outside the soft-reset loop.

Running the update inside the soft-reset loop will mean that (on boards
like PYBD that use a bootloader) the same reset mode is used each
reset loop, eg factory reset occurs each time.

Signed-off-by: Damien George <damien@micropython.org>

iabdalkader committed:
stm32/sdio: Don't change any DMA2 settings on H7 MCUs.

DMA2 clock and registers should be left in their current state in the H7
build.

iabdalkader committed:
stm32/sdcard: Fix H7 build when using SDMMC2.

Changes are:
- Fix missing IRQ handler when SDMMC2 is used instead of SDMMC1 with H7
  MCUs.
- Removed outdated H7 series compatibility macros.
- Defined common IRQ handler macro for F4 series.

esmil committed:
py/scope: Name and use id_kind_type_t.

The function scope_find_or_add_id used to take a scope_kind_t enum and
save it in an uint8_t. Saving an enum in a uint8_t is fine, but
everywhere this function is called it is not actually given a
scope_kind_t but an anonymous enum instead. Let's give this enum a name
and use that as the argument type.

This doesn't change the generated code, but is a C type mismatch that
unfortunately doesn't show up unless you enable -Wenum-conversion.

esmil committed:
py: Use unsigned comparison of chars.

On x86 chars are signed, but we're comparing a char to '0' + unsigned int,
which is promoted to an unsigned int. Let's promote the char to unsigned
before doing the comparison to avoid weird corner cases.

esmil committed:
py, extmod: Add explicit initializers for default values.

When compiling with -Wextra which includes -Wmissing-field-initializers
GCC will warn that the defval field of mp_arg_val_t is not initialized.
This is just a warning as it is defined to be zero initialized, but since
it is a union it makes sense to be explicit about which member we're
going to use, so add the explicit initializers and get rid of the
warning.

esmil committed:
mpy-cross: Enable more warnings.

esmil committed:
unix: Enable more warnings.

esmil committed:
extmod: Disable -Wmissing-field-initializers for lfs2.

mytechnotalent committed:
docs/reference/glossary.rst: Fix minor grammar error, An -> A.

Commit:
tools/upip.py: Support explicit port number in host.

Adding a port number other then 443 to a PyPI URL may be needed if a local
server like devpi is used.

stinos committed:
py/mkrules.mk: Add target for compiling C++ files.

Move the target from the ESP32 Makefile since that does what is needed
already, but also include files from user C modules as is done for the C
files.

stinos committed:
docs: Fix reference to QSTR_GEN_CFLAGS Makefile flag.

stinos committed:
py/makeqstrdefs.py: Support preprocessing C++ files for QSTR generation.

When SCR_QSTR contains C++ files they should be preprocessed with the same
compiler flags (CXXFLAGS) as they will be compiled with, to make sure code
scanned for QSTR occurrences is effectively the code used in the rest of
the build.  The 'split SCR_QSTR in .c and .cpp files and process each with
different flags' logic isn't trivial to express in a Makefile and the
existing principle for deciding which files to preprocess was already
rather complicated, so the actual preprocessing is moved into
makeqstrdefs.py completely.

stinos committed:
py/py.mk: Support C++ code for user C modules.

Support C++ code in .cpp files by providing CXX counterparts of the
_USERMOD_ flags we have for C already.  This merely enables the Makefile of
user C modules to use variables specific to C++ compilation, it is still up
to each port's main Makefile to also include these in the build.

stinos committed:
esp32,unix: Support building C++ code.

Support building .cpp files and linking them into the micropython
executable in a way similar to how it is done for .c files.  The main
incentive here is to enable user C modules to use C++ files (which are put
in SRC_MOD_CXX by py.mk) since the core itself does not utilize C++.

However, to verify build functionality a unix overage test is added.  The
esp32 port already has CXXFLAGS so just add the user modules' flags to it.
For the unix port use a copy of the CFLAGS but strip the ones which are not
usable for C++.

jimmo committed:
stm32/rfcore: Update to support WS=1.9.0.0.4.

This WS update to 1.9.0.0.4 broke the workaround used in rfcore for
OCF_CB_SET_EVENT_MASK2, so fix it to support WS 1.8 and 1.9.

jimmo committed:
py/objdict: Add mp_const_empty_dict_obj, use it for mp_const_empty_map.

hlovatt committed:
docs/library/btree.rst: Correct method typo: __detitem__ to __delitem__.

iabdalkader committed:
stm32: Fix broken build when FAT FS multi-partition is disabled.

hlovatt committed:
docs/library/machine.rst: Correct minor typo: timout to timeout.

hlovatt committed:
docs/library/machine.Signal.rst: Correct typo: usecases to use cases.

hlovatt committed:
docs/library/machine.Timer.rst: Add mention of constructor arguments.

dpgeorge committed:
ports: Use correct in/out endpoint size in TUD_CDC_DESCRIPTOR.

The last argument of TUD_CDC_DESCRIPTOR() is the endpoint size (or
wMaxPacketSize), not the CDC RX buffer size (which can be larger than the
endpoint size).

Signed-off-by: Damien George <damien@micropython.org>

stinos committed:
py: Workaround clang error when building misc.h with C++ compiler.

jimmo committed:
stm32/boards/NUCLEO_WB55/rfcore_firmware.py: Fix flash unlock.

The flash can sometimes be in an already-unlocked state, and attempting to
unlock it again will cause an immediate reset.  So make _Flash.unlock()
check FLASH_CR_LOCK to get the current state.

Also fix some magic numbers for FLASH_CR_LOCK AND FLASH_CR_STRT.

The machine.reset() could be removed because it no longer crashes now that
the flash unlock is fixed.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32/boards/NUCLEO_WB55/rfcore_firmware.py: Fix bad variable name.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32/boards/NUCLEO_WB55/rfcore_firmware.py: Increase GET_STATE timeout.

When installing WS firmware, the very first GET_STATE can take several
seconds to respond (especially with the larger binaries like
BLE_stack_full).

Allows stm.rfcore_sys_hci to take an optional timeout, defaulting to
SYS_ACK_TIMEOUT_MS (which is 250ms).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

pi-anl committed:
unix: Handle pending events/scheduler in MICROPY_EVENT_POLL_HOOK.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge committed:
stm32/rng: Use SysTick+RTC+unique-id to seed pRNG for MCUs without RNG.

The same seed will only occur if the board is the same, the RTC has the
same time (eg freshly powered up) and the first call to this function (eg
via an "import random") is done at exactly the same time since reset.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/modurandom: Support urandom.seed() without an argument.

If a port provides MICROPY_PY_URANDOM_SEED_INIT_FUNC as a source of
randomness then this will be used when urandom.seed() is called without
an argument (or with None as the argument) to seed the pRNG.

Other related changes in this commit:
- mod_urandom___init__ is changed to call seed() without arguments, instead
  of explicitly passing in the result of MICROPY_PY_URANDOM_SEED_INIT_FUNC.
- mod_urandom___init__ will only ever seed the pRNG once (before it could
  seed it again if imported by, eg, random and then urandom).
- The Yasmarang state is moved to the BSS for builds where the state is
  guaranteed to be initialised on import of the (u)random module.

Signed-off-by: Damien George <damien@micropython.org>

andrewleech committed:
esp32/mpconfigport.h: Enable MICROPY_PY_DELATTR_SETATTR.

To align with unix and stm32 ports.

stinos committed:
examples: Add example code for user C modules, both C and C++.

Add working example code to provide a starting point for users with files
that they can just copy, and include the modules in the coverage test to
verify the complete user C module build functionality.  The cexample module
uses the code originally found in cmodules.rst, which has been updated to
reflect this and partially rewritten with more complete information.

dpgeorge committed:
stm32/usbd_cdc_interface: Check and handle CDC TX wrap-overflow.

If the device is not connected over USB CDC to a host then all output to
the CDC (eg initial boot messages) is written to the CDC TX buffer with
wrapping, so that the most recent data is retained when the USB CDC is
eventually connected (eg so the REPL banner is displayed upon connection).

This commit fixes a bug in this behaviour, which was likely introduced in
e4fcd216e02eef0b389c84ecd67be3114aac0a5d, where the initial data in the CDC
TX buffer is repeated multiple times on first connection of the device to
the host.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Thanks for the contribution!

I think this change is good, but I think would be better if it were in the main `ports/nrf/README.md` file because it could relate to other boards with more than GPIO port.

stinos committed:
py/makeqstrdefs.py: Process C++ files as well.

Preprocessed C++ code isn't different from C code when it comes to QSTR
instances so process it as well.

JPFrancoia commented:
Sure, shouldn't be a problem, but can this numbering scheme be applied to all boards?

dpgeorge committed:
stm32: Add MICROPY_BOARD calls in various places in stm32_main.

For a board to have full configurability of the soft reset loop.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/thread/stress_schedule.py: Assign globals before running test.

When threading is enabled without the GIL then there can be races between
the threads accessing the globals dict.  Avoid this issue by making sure
all globals variables are allocated before starting the threads.

Signed-off-by: Damien George <damien@micropython.org>

jimmo committed:
extmod/modbluetooth: Re-instate optional no-ringbuf modbluetooth.

This requires that the event handlers are called from non-interrupt context
(i.e. the MicroPython scheduler).

This will allow the BLE stack (e.g. NimBLE) to run from the scheduler
rather than an IRQ like PENDSV, and therefore be able to invoke Python
callbacks directly/synchronously.  This allows writing Python BLE handlers
for events that require immediate response such as _IRQ_READ_REQUEST (which
was previous a hard IRQ) and future events relating to pairing/bonding.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Make UUID type accessible outside modbluetooth.c.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge committed:
stm32/machine_adc: Fix ADC auto-calibration to run when ADC not enabled.

Prior to this commit, the ADC calibration code was never executing because
ADVREGEN bit was set making the CR register always non-zero.

This commit changes the logic so that ADC calibration is always run when
the ADC is disabled and an ADC channel is initialised.  It also uses the LL
API functions to do the calibration, to make sure it is done correctly on
each MCU variant.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/vfs_lfs: Support mounting LFS filesystems in read-only mode.

Signed-off-by: Damien George <damien@micropython.org>

jimmo committed:
extmod/btstack/btstack.mk: Add -Wimplicit-fallthrough=0.

This is needed since -Wextra was added to the build in
bef412789ea93c521bd9c2dddc22b9b3484da574

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/makeqstrdefs.py: Run qstr preprocessing in parallel.

This gives a substantial speedup of the preprocessing step, i.e. the
generation of qstr.i.last.  For example on a clean build, making
qstr.i.last:

    21s -> 4s on STM32 (WB55)
    8.9 -> 1.8s on Unix (dev).

Done in collaboration with @stinos.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge committed:
docs/develop/cmodules.rst: Add link to source code for user C example.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
unix/Makefile: Move coverage.c and coveragecpp.cpp to coverage variant.

So that g++ is not needed to build a non-coverage unix variant.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32: Support C++ code and user C modules written in C++.

Also build user C modules as part of the stm32 CI.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/micropython/extreme_exc.py: Unlink alloc'd lists earlier in chain.

To help the GC collect this memory that's no longer needed after the test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/boards: Factor out common data/bss/heap/stack linker sections.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
lib/utils/pyexec: Add MICROPY_BOARD hooks before/after executing code.

Signed-off-by: Damien George <damien@micropython.org>

Arrowana committed:
extmod/machine_mem: Only allow integers in machine.memX subscript.

Prior to this change machine.mem32['foo'] (or using any other non-integer
subscript) could result in a fault due to 'foo' being interpreted as an
integer.  And when writing code it's hard to tell if the fault is due to a
bad subscript type, or an integer subscript that specifies an invalid
memory address.

The type of the object used in the subscript is now tested to be an
integer by using mp_obj_get_int_truncated instead of
mp_obj_int_get_truncated.  The performance hit of this change is minimal,
and machine.memX objects are more for convenience than performance (there
are many other ways to read/write memory in a faster way),

Fixes issue #6588.

jimmo committed:
tests/run-multitests.py: Fix diff order, show changes relative to truth.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth: Improve reliability of event waiting.

Use the same `wait_for_event` in all tests that doesn't hold a reference to
the event data tuple and handles repeat events.

Also fix a few misc reliability issues around timeouts and sequencing.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/run-multitests.py: Add a -p flag to run permutations of instances.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Combine gattc-data-available callbacks into one.

Instead of having the stack indicate a "start", "data"..., "end", pass
through the data in one callback as an array of chunks of data.

This is because the upcoming non-ringbuffer modbluetooth implementation
cannot buffer the data in the ringbuffer and requires instead a single
callback with all the data, to pass to the Python callback.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Poll startup directly rather than using NimBLE sem.

Using a semaphore (the previous approach) will only run the UART, whereas
for startup we need to also run the event queue.

This change makes it run the full scheduler hook.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
unix: Make mp_hal_delay_ms run MICROPY_EVENT_POLL_HOOK.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Make stm32 and unix NimBLE ports use synchronous events.

This changes stm32 from using PENDSV to run NimBLE to use the MicroPython
scheduler instead.  This allows Python BLE callbacks to be invoked directly
(and therefore synchronously) rather than via the ringbuffer.

The NimBLE UART HCI and event processing now happens in a scheduled task
every 128ms.  When RX IRQ idle events arrive, it will also schedule this
task to improve latency.

There is a similar change for the unix port where the background thread now
queues the scheduled task.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble/nimble.mk: Add -Wno-old-style-declaration.

This is needed since -Wextra was added to the build in
bef412789ea93c521bd9c2dddc22b9b3484da574

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

nedseb committed:
stm32/Makefile: Make the generation of `firmware.bin` explicit.

The file `$(BUILD)/firmware.bin` was used by the target `deploy-stlink` and
`deploy-openocd` but it was generated indirectly by the target
`firmware.dfu`.

As this file could be used to program boards directly by a Mass Storage
copy, it's better to make it explicitly generated.

Additionally, some target are refactored to remove redundancy and be more
explicit on dependencies.

dpgeorge committed:
py/mpz: Do sign extension in mpz_as_bytes for negative values.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/moductypes: Fix storing to (U)INT64 arrays on 32-bit archs.

Fixes issue #6583.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
py/binary: Fix sign extension setting wide integer on 32-bit archs.

Signed-off-by: Damien George <damien@micropython.org>

jonathanhogg committed:
esp32/machine_pin: Reset pin if init sets mode.

This will forcibly grab the pin back from the ADC if it has previously been
associated with it.

Fixes #5771.

dpgeorge committed:
stm32/boardctrl: Define MICROPY_BOARD_EARLY_INIT alongside others.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/Makefile: Allow boards to extend all SRC variables.

And rename SRC_HAL -> HAL_SRC_C and SRC_USBDEV -> USBDEV_SRC_C for
consistency with other source variables.

Follow on from 0fff2e03fe07471997a6df6f92c6960cfd225dc0

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/rtc: Validate the RTC prescaler on boot and change if incorrect.

Devices with RTC backup-batteries have been shown (very rarely) to have
incorrect RTC prescaler values.  Such incorrect values mean the RTC counts
fast or slow, and will be wrong forever if the power/backup-battery is
always present.

This commit detects such a state at start up (hard reset) and corrects it
by reconfiguring the RTC prescaler values.

Signed-off-by: Damien George <damien@micropython.org>

glennrub commented:
I also believe it could be made generic as @dpgeorge suggest (`ports/nrf/README.md`). There are three nrf52840 boards in the port, and more boards might be added in the future.

glennrub commented on ports/nrf/boards/pca10059/readme.md:
```diff
@@ -0,0 +1,7 @@
+# Pins mapping
+
+Software Pins 0-31 correspond to physical pins 0.x and software Pins 32-63
```
I would suggest to change "... software Pins 32-63" to "... software Pins 32-48" as nrf52840 only provides 48 GPIO's. (The same is the case for the nrf53 device which also have 48 GPIOs).

JPFrancoia replied:
Done.

glennrub replied:
I'm so sorry!
Terrible mistake from my side. Of course should would be 32 - **47**, not 48.

JPFrancoia replied:
No problem. I addressed it

JPFrancoia committed:
Merge branch 'master' of github.com:JPFrancoia/micropython

JPFrancoia committed:
Movinf pins mapping documentation for nrf52840 boards to /nrf/README.md,
to make it more generic.

JPFrancoia committed:
Addressing max pin number for nrf numbering scheme.

glennrub commented:
Thanks @JPFrancoia , rebased, squshed and merged in 3dcb551d8981bfb370bfe8f467fcde59dd7a916a.
